### PR TITLE
DNM: kraken: osd: calc_clone_subsets misuses try_read_lock vs missing

### PR DESCRIPTION
**Pull request by shinobu-x**
http://tracker.ceph.com/issues/18723

### CODE REVIEW

**Comment by shinobu-x** [comment]
Another storage build failure here...

**Comment by shinobu-x** [comment]
Should work hopefully!

**Comment by shinobu-x** [comment]
=)

**Comment by shinobu-x** [comment]
jenkins test this please (no logs)

**Comment by shinobu-x** [comment]
Getting excited...

**Comment by tchaikov** [comment]
@shinobu-x you need to backport the fix for http://tracker.ceph.com/issues/17831 as well, i.e. the changes you are backporting here is based on #12844.

**Comment by smithfarm** [comment]
@shinobu-x Also, could you please edit the description of the first commit to explain what the conflict in `src/osd/PrimaryLogPG.h` was and how you resolved it?

**Comment by shinobu-x** [comment]
@tchaikov @smithfarm Got you, guys :)

**Comment by shinobu-x** [comment]
@tchaikov Cherry-picked suggested commit.

**Comment by shinobu-x** [comment]
@smithfarm Here is what I did:
First off I fixed conflict with normal way.

After double-checking, I noticed that there were some lines which were not included picked file. I added them (See the following log).

If you think that it's better to revert,  I will.

==== Log ====
diff --git a/src/osd/PrimaryLogPG.h b/src/osd/PrimaryLogPG.h
index fb67087..02ab041 100644
--- a/src/osd/PrimaryLogPG.h
+++ b/src/osd/PrimaryLogPG.h
@@ -335,6 +335,7 @@ public:
   const pg_pool_t &get_pool() const override {
     return pool.info;
   }
+
   ObjectContextRef get_obc(
     const hobject_t &hoid,
     map<string, bufferlist> &attrs) override {
@@ -347,6 +348,10 @@ public:
     ObcLockManager &manager) override {
     if (is_missing_object(hoid))
       return false;
+
+  bool try_lock_for_read(
+    const hobject_t &hoid,
+    ObcLockManager &manager) override {
     auto obc = get_object_context(hoid, false, nullptr);
     if (!obc)
       return false;
@@ -369,7 +374,6 @@ public:
     return clear_object_snap_mapping(t, soid);
   }
 
-
   void log_operation(
     const vector<pg_log_entry_t> &logv,
     boost::optional<pg_hit_set_history_t> &hset_history,

**Comment by smithfarm** [comment]
@shinobu-x The difference is just whitespace? Can you use `git blame` on master to find the commit that added those blank lines?

**Comment by shinobu-x** [comment]
@smithfarm These lines were missed, first time I committed.

> +  bool try_lock_for_read(
> +    const hobject_t &hoid,
> +    ObcLockManager &manager) override {

File: src/osd/PrimaryLogPG.h [hunk 1]
```diff
@@ -347,6 +348,10 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
     ObcLockManager &manager) override {
     if (is_missing_object(hoid))
       return false;
+
+  bool try_lock_for_read(
+    const hobject_t &hoid,
+    ObcLockManager &manager) override {
```
Comment (smithfarm) [inline src/osd/PrimaryLogPG.h]: @shinobu-x This isn't good. These three lines you added are repeated 5 lines up.

Comment (shinobu-x) [inline src/osd/PrimaryLogPG.h]: @smithfarm Got it =)

**Pull request closed without merging**